pkg/v21/resource/configmapmigration/create.go

```diff
@@ -112,6 +116,68 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		return microerror.Mask(err)
 	}
 
+	for _, chartSpec := range chartSpecsToMigrate {
+		if chartSpec.UserConfigMapName != "" {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding out if user configmap %#q has been migrated", chartSpec.UserConfigMapName))
+
+			cm, err := getConfigMapByName(clusterConfigMaps.Items, chartSpec.UserConfigMapName)
+			if IsNotFound(err) {
+				// Copy user configmap from the tenant cluster to the cluster namespace.
+				err = r.copyUserConfigMap(ctx, cc.Client.TenantCluster.K8s, cr, chartSpec)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("user configmap %#q has been migrated", chartSpec.UserConfigMapName))
+			} else if cm.Name == chartSpec.UserConfigMapName {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("user configmap %#q has already been migrated", chartSpec.UserConfigMapName))
+			} else if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) copyUserConfigMap(ctx context.Context, tenantK8sClient kubernetes.Interface, cr v1alpha1.ClusterGuestConfig, chartSpec key.ChartSpec) error {
+	currentCM, err := tenantK8sClient.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(chartSpec.UserConfigMapName, metav1.GetOptions{})
+	if IsNotFound(err) || len(currentCM.Data) == 0 {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("user configmap %#q has no data to migrate", chartSpec.UserConfigMapName))
+		return nil
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating user configmap %#q in namespace %#q", chartSpec.UserConfigMapName, key.ClusterID(cr)))
+
+	// User configmaps for chartconfig CRs only have keys and values under the
+	// configmap block. This needs to be converted to YAML for app CRs.
+	values := map[string]interface{}{
+		"configmap": currentCM.Data,
+	}
+
+	yamlValues, err := yaml.Marshal(values)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	cm := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      chartSpec.UserConfigMapName,
+			Namespace: key.ClusterID(cr),
```

Comment:
Are we certain this namespace exists? AFAIK older clusters on AWS and Azure might not have such a namespace.

Reply:
Since this is only v21 changes, I don't think we would get a problem from older clusters.

Reply:
I am not sure how the migration works when a cluster is upgraded. If a namespace is created then this is fine.

Reply:
Like Ji says this is only for v21 clusters. Also IIRC all tenants should have this because its used by the endpoints resource which is needed for scraping tenant clusters.

Reply:
> I am not sure how the migration works when a cluster is upgraded.

@MarcelMue Thanks I'll check. We also use this namespace for the app CRs.

Ideally this namespace should be created by cluster-operator and its great that was done for the clusterapi controller.